lms/services/blackboard_api/client.py

```diff
@@ -88,10 +78,74 @@ def list_files(self, course_id, folder_id=None):
                     "size": file["size"],
                     "parent_lms_id": file["parentId"],
                 }
-                for file in results
+                for file in files
             ]
         )
 
+    def list_files(self, course_id, folder_id=None):
+        """Return the list of files in the given course or folder."""
+        results = []
+
+        path = self._list_files_url(course_id, folder_id)
+
+        for _ in range(PAGINATION_MAX_REQUESTS):
+            response = self._api.request("GET", path)
+            results.extend(BlackboardListFilesSchema(response).parse())
+            path = response.json().get("paging", {}).get("nextPage")
+            if not path:
+                break
+
+        self._store_files(course_id, results)
+
+        return results
+
+    def _get_all_folders_files(self, course_id, results):
+        # Get all the folder URL from `results`
+        folder_urls = [
+            self._api._api_url(self._list_files_url(course_id, result["id"]))
+            for result in results
+            if result["type"] == "Folder"
+        ]
+
+        # Query all folders at once
+        responses = self._request.find_service(name="async_oauth_http").request(
+            "GET", folder_urls
+        )
+
+        for response in responses:
+            # Get the first page result
+            folder_results = BlackboardListFilesSchema(response).parse()
+            # Add any extra pages
+            for _ in range(PAGINATION_MAX_REQUESTS - 1):
+                path = response.sync_json.get("paging", {}).get("nextPage")
+                if not path:
+                    break
+                response = self._api.request("GET", path)
+                folder_results.extend(BlackboardListFilesSchema(response).parse())
+
+            # For each folder, get it children too
+            results.extend(self._get_all_folders_files(course_id, folder_results))
+
+        return results
+
+    def list_all_files(self, course_id):
```

Comment:
This has a lot of copy pasted code but the problem is trying to solve is:


- We need all the files in the new course to be able to fix course copy.
- In BB we only get files as the user navigates over the folders.


We'd use this method only during course copy to get the list of files of the new course.

We could try to use the stored info for the original file to only query the folder structure of that tree. 

We'll save some API queries but:

- That code will be more complex, we need to match the tree structure of both courses
- It will need to be executed once per assignment, as opposed to once per course of this version.


Some light checking of the info we have in the DB suggest that the worst offenders are folders with many pages not very deep folder structures:

https://hypothes-is.slack.com/archives/C04JN7PMG9X/p1674234224864519